File: src/main/java/com/nobot/plugin/TodayLuck.java
package com.nobot.plugin;

import com.IceCreamQAQ.Yu.annotation.Action;
import com.IceCreamQAQ.Yu.annotation.Before;
import com.icecreamqaq.yuq.annotation.GroupController;
import com.icecreamqaq.yuq.annotation.PrivateController;
import com.icecreamqaq.yuq.controller.BotActionContext;
import com.icecreamqaq.yuq.entity.Contact;
import com.icecreamqaq.yuq.entity.Friend;
import com.icecreamqaq.yuq.entity.Group;
import com.icecreamqaq.yuq.entity.Member;
import com.icecreamqaq.yuq.message.Message;
import com.icecreamqaq.yuq.message.MessageLineQ;
import com.nobot.tool.MD5Utils;
import com.nobot.tool.TimeUtils;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

@GroupController
@PrivateController
public class TodayLuck
{
	private final SimpleDateFormat simpleDateFormat=new SimpleDateFormat("yyyyMMdd");
	@Before
	public void getName(BotActionContext actionContext, Contact qq)
	{
		if(actionContext.getSource() instanceof Member)
		{
			actionContext.set("name",qq.getName());
			return;
		}
		if(qq instanceof Friend)
		{
			actionContext.set("name",qq.getName());
			return;
		}
		String name=((Member) qq).getNameCard();
		if(name.isEmpty())
		{
			name=qq.getName();
		}
		actionContext.set("name",name);
	}

	@Action("今日人品")
	public MessageLineQ getTodayLuck(long qq,String name)
	{
		MessageLineQ messageLineQ=new Message().lineQ();
		long nowTime=System.currentTimeMillis();
		Date today=new Date(nowTime);
		Date yesterday;
		Calendar calendar=Calendar.getInstance();
		calendar.setTime(today);
		calendar.add(Calendar.DATE,-1);
		yesterday=calendar.getTime();
		String todayInfo= TimeUtils.getFormatData(simpleDateFormat)+name;
		String todayResult=MD5Utils.getMD5(todayInfo,10);

		int index=Integer.parseInt(todayResult.substring(1,2));
		int todayLuck=Integer.parseInt(todayResult.substring(index+2,index+4));
		String yesterdayInfo=simpleDateFormat.format(yesterday);
		String yesterdayResult=MD5Utils.getMD5(yesterdayInfo,10);

		index=Integer.parseInt(todayResult.substring(1,2));
		int yesterdayLuck=Integer.parseInt(yesterdayResult.substring(index+2,index+4));
		int different=Math.round((todayLuck-yesterdayLuck)/20f);
		messageLineQ.text("今天"+name+"的运气是"+todayLuck+"\r\n");
		switch (different)
		{
			case 5:
				messageLineQ.text("时来运转！运气的惊天逆转，从昨日低谷瞬间冲顶。昨日的枷锁今天一甩而开吧。");
				break;
			case 4:
				messageLineQ.text("幸运upup！相比昨天可是地下天上。全新的生活等着你！");
				break;
			case 3:
				if(yesterdayLuck<30)
					messageLineQ.text("运气大幅提升，想必昨日的困难今天将会结束。");
				else
					messageLineQ.text("很棒哦！一部登顶的感觉很爽，就这这个劲头继续吧。");
				break;
			case 2:
				if (yesterdayLuck<30)
					messageLineQ.text("虽然昨天不太舒服，不过今天有了很大的起色哦。");
				else
					messageLineQ.text("哇 今天比昨天要顺利的多呢。");
				break;
			case 1:
				if (yesterdayLuck<30)
					messageLineQ.text("缓慢的进展也是好的。");
				else if(yesterdayLuck<60)
					messageLineQ.text("运气在上升哦！");
				else
					messageLineQ.text("更进一步！势头很棒，就着这种势头前进吧。");
				break;
			case 0:
				if (yesterdayLuck<30)
					messageLineQ.text("似乎在低谷徘徊呢，不要气馁，相信自己。");
				else if(yesterdayLuck<60)
					messageLineQ.text("不以物喜，不以己悲，依靠自己的力量。");
				else
					messageLineQ.text("一鼓作气，干柴烈火，趁机多干点自己的宏伟计划吧。");
				break;
			case -1:
				if (todayLuck<30)
					messageLineQ.text("不要太难过了……运由天注定，命靠自己打拼。");
				else if(todayLuck<60)
					messageLineQ.text("微微的下降不要担心，相信运气会跟着你的努力回来的。");
				else
					messageLineQ.text("盛极必衰，建议韬光养晦。");
				break;
			case -2:
				if (todayLuck<30)
					messageLineQ.text("加油啊！运气不会决定一切。");
				else
					messageLineQ.text("运气掉了不少吗？别总是依赖运气啊。");
				break;
			case -3:
				if(todayLuck<30)
					messageLineQ.text("啊 有时确实会有这样的情况嘛，毕竟运气不会久留。");
				else
					messageLineQ.text("还没触底，还没触底，还有机会，还有机会。");
				break;
			case -4:
				messageLineQ.text("不要太难过了啊，事情不会那么糟的。");
				break;
			case -5:
				messageLineQ.text("有一说一 你这样运气过山车很罕见啊，从某种情况来说算不算是一种运气？");
				break;
		}
		messageLineQ.text("\r\n");
		messageLineQ.text("今天好方向：");
		int place= (int) (Integer.parseInt(todayResult.substring(12,13))/2.5);
		switch (place)
		{
			case 0:
				messageLineQ.text("东");
				break;
			case 1:
				messageLineQ.text("西");
				break;
			case 2:
				messageLineQ.text("南");
				break;
			case 3:
				messageLineQ.text("北");
				break;
		}
		messageLineQ.text("\t");
		messageLineQ.text("今天好颜色：");
		int color=Integer.parseInt(todayResult.substring(13,14));
		switch (place)
		{
			case 0:
				messageLineQ.text("红");
				break;
			case 1:
				messageLineQ.text("橙");
				break;
			case 2:
				messageLineQ.text("黄");
				break;
			case 3:
				messageLineQ.text("绿");
				break;
			case 4:
				messageLineQ.text("青");
				break;
			case 5:
				messageLineQ.text("蓝");
				break;
			case 6:
				messageLineQ.text("紫");
				break;
			case 7:
				messageLineQ.text("黑");
				break;
			case 8:
				messageLineQ.text("白");
				break;
			case 9:
				messageLineQ.text("粉");
				break;
		}

		return messageLineQ;
	}
}

File: src/main/java/com/nobot/plugin/TodayLuck.kt
package com.nobot.plugin

import com.IceCreamQAQ.Yu.annotation.Action
import com.IceCreamQAQ.Yu.annotation.Before
import com.icecreamqaq.yuq.annotation.GroupController
import com.icecreamqaq.yuq.annotation.PrivateController
import com.icecreamqaq.yuq.controller.BotActionContext
import com.icecreamqaq.yuq.entity.Contact
import com.icecreamqaq.yuq.entity.Friend
import com.icecreamqaq.yuq.entity.Group
import com.icecreamqaq.yuq.entity.Member
import com.icecreamqaq.yuq.message.Message
import com.icecreamqaq.yuq.message.MessageLineQ
import com.nobot.system.annotation.UnzipFile
import com.nobot.system.stringHelper.DefaultStringFile
import com.nobot.system.stringHelper.GetString
import com.nobot.tool.MD5Utils
import com.nobot.tool.TimeUtils
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject
import kotlin.math.roundToInt

@GroupController
@PrivateController
@UnzipFile(name = "string/TodayLuck.properties", aim = "./string/TodayLuck.properties")
@DefaultStringFile(value = "./String/TodayLuck.properties", info = "这是今日人品的相关文件")
class TodayLuck
{
	private val simpleDateFormat = SimpleDateFormat("yyyyMMdd")
	@Inject
	private lateinit var getString:GetString

	@Before
	fun getName(actionContext: BotActionContext, qq: Contact,group: Group?)
	{
		if (actionContext.source is Member)
		{
			actionContext["name"] = qq.name
			return
		}
		if (qq is Friend)
		{
			actionContext["name"] = qq.name
			return
		}
		var name = (qq as Member).nameCard
		if (name.isEmpty())
		{
			name = qq.name
		}
		actionContext["name"] = name
		actionContext["groupName"]=group?.name ?:""

	}

	@Action("今日人品")
	fun getTodayLuck(qq: Long, name: String,groupName:String): MessageLineQ
	{
		val messageLineQ = Message().lineQ()
		val nowTime = System.currentTimeMillis()
		val today = Date(nowTime)
		val yesterday: Date
		val calendar = Calendar.getInstance()
		calendar.time = today
		calendar.add(Calendar.DATE, -1)
		yesterday = calendar.time
		val todayInfo = TimeUtils.getFormatData(simpleDateFormat) + name
		val todayResult = MD5Utils.getMD5(todayInfo, 10)
		var index = todayResult.substring(1, 2).toInt()
		val todayLuck = todayResult.substring(index + 2, index + 4).toInt()
		val yesterdayInfo = simpleDateFormat.format(yesterday)
		val yesterdayResult = MD5Utils.getMD5(yesterdayInfo, 10)
		index = todayResult.substring(1, 2).toInt()
		val yesterdayLuck = yesterdayResult.substring(index + 2, index + 4).toInt()
		val different = ((todayLuck - yesterdayLuck) / 20f).roundToInt()

		val args= arrayOfNulls<String>(4)
		args[0]=name
		args[1]=groupName
		args[2]= todayLuck.toString()
		args[3]=yesterdayLuck.toString()

		messageLineQ.text(getString.formatString(getString.addressing("jrrp.start"), *args))
		var address= "jrrp.${different}."
		when (different)
		{
			3 -> address += if (yesterdayLuck < 20) '2' else '4'
			2 -> address += if (yesterdayLuck < 30) '3' else '6'
			1 -> address += when(yesterdayLuck)
			{
				in 0..20->2
				in 0..50->5
				in 0..80->8
				else ->8
			}
			0 -> address += when(yesterdayLuck)
			{
				in 0..30->3
				in 0..60->6
				in 0..100->10
				else ->10
			}
			-1 -> address += when(yesterdayLuck)
			{
				in 0..50->5
				in 0..80->8
				in 0..100->10
				else ->8
			}
			-2 -> address += if (yesterdayLuck < 70) '7' else "10"
			-3 -> address += if (yesterdayLuck < 80) '8' else "10"
			else -> address.dropLast(1)
		}
		messageLineQ.text(getString.formatString(getString.addressing(address),*args))
		messageLineQ.text("\r\n")
		messageLineQ.text("今天好方向：")
		messageLineQ.text(when ((todayResult.substring(12, 13).toInt() / 2.5).toInt())
		{
			0 -> "东"
			1 -> "西"
			2 -> "南"
			3 -> "北"
			else ->""
		})
		messageLineQ.text("\t")
		messageLineQ.text("今天好颜色：")
		messageLineQ.text(when (todayResult.substring(13, 14).toInt())
		{
			0 -> "红"
			1 -> "橙"
			2 -> "黄"
			3 -> "绿"
			4 -> "青"
			5 -> "蓝"
			6 -> "紫"
			7 -> "黑"
			8 -> "白"
			9 -> "粉"
			else -> ""
		})
		return messageLineQ
	}
}